src/test/stateful_modelinstance_test.cpp

```diff
@@ -165,6 +167,74 @@ class StatefulModelInstanceTest : public ::testing::Test {
     }
 };
 
+void StartStatefulPredict(const std::shared_ptr<ovms::ModelInstance> modelInstance, inputs_info_t modelInput, uint64_t seqId) {
+    tensorflow::serving::PredictRequest request = preparePredictRequest(modelInput);
+    setRequestSequenceId(&request, seqId);
+    setRequestSequenceControl(&request, SEQUENCE_START);
+
+    std::unique_ptr<ovms::ModelInstanceUnloadGuard> unload_guard;
+    tensorflow::serving::PredictResponse response;
+    // Do the inference
+    ASSERT_EQ(modelInstance->infer(&request, &response, unload_guard), ovms::StatusCode::OK);
+    // Check response
+    EXPECT_TRUE(CheckSequenceIdResponse(response, seqId));
+}
+
+void EndStatefulPredict(const std::shared_ptr<ovms::ModelInstance> modelInstance, inputs_info_t modelInput, uint64_t seqId) {
+    tensorflow::serving::PredictRequest request = preparePredictRequest(modelInput);
+    setRequestSequenceId(&request, seqId);
+    setRequestSequenceControl(&request, SEQUENCE_END);
+
+    std::unique_ptr<ovms::ModelInstanceUnloadGuard> unload_guard;
+    tensorflow::serving::PredictResponse response;
+    // Do the inference
+    ASSERT_EQ(modelInstance->infer(&request, &response, unload_guard), ovms::StatusCode::OK);
+    // Check response
+    EXPECT_TRUE(CheckSequenceIdResponse(response, seqId));
+}
+
+void RunStatefulPredicts(const std::shared_ptr<ovms::ModelInstance> modelInstance, inputs_info_t modelInput, int numberOfNoControlRequests, uint64_t seqId,
+    std::unique_ptr<std::future<void>> waitAfterSequenceStarted,
+    std::unique_ptr<std::future<void>> waitBeforeSequenceFinished) {
+    std::unique_ptr<ovms::ModelInstanceUnloadGuard> unload_guard;
+
+    std::vector<tensorflow::serving::PredictRequest> requests;
+    std::vector<tensorflow::serving::PredictResponse> responses;
+    for (int i = 0; i < numberOfNoControlRequests; i++) {
+        tensorflow::serving::PredictRequest request = preparePredictRequest(modelInput);
+        requests.push_back(request);
+    }
+
+    for (int i = 0; i < numberOfNoControlRequests; i++) {
+        tensorflow::serving::PredictResponse response;
+        responses.push_back(response);
+    }
+
+    StartStatefulPredict(modelInstance, modelInput, seqId);
+    if (waitAfterSequenceStarted) {
+        std::cout << "Waiting after StartStatefulPredict" << std::endl;
+        waitAfterSequenceStarted->get();
+    }
+    for (int i = 1; i < numberOfNoControlRequests; i++) {
+        tensorflow::serving::PredictRequest request = requests[i];
```

Comment:
If we use and check one request at a time in this loop we can create request here and it will reside in the scope of particular iteration of the loop. So you could change this line to:
` tensorflow::serving::PredictRequest request = preparePredictRequest(modelInput);`